coderd/workspaceapps/stats.go

```diff
@@ -0,0 +1,196 @@
+package workspaceapps
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/codersdk"
+)
+
+const (
+	DefaultStatsCollectorReportInterval = 30 * time.Second
+)
+
+// StatsReport is a report of a workspace app session.
+type StatsReport struct {
+	UserID           uuid.UUID         `json:"user_id"`
+	WorkspaceID      uuid.UUID         `json:"workspace_id"`
+	AgentID          uuid.UUID         `json:"agent_id"`
+	AccessMethod     AccessMethod      `json:"access_method"`
+	SlugOrPort       string            `json:"slug"`
+	SessionID        uuid.UUID         `json:"session_id"`
+	SessionStartTime time.Time         `json:"session_start_time"`
+	SessionEndTime   codersdk.NullTime `json:"session_end_time"`
+}
+
+func newStatsReportFromSignedToken(token SignedToken) StatsReport {
+	return StatsReport{
+		UserID:           token.UserID,
+		WorkspaceID:      token.WorkspaceID,
+		AgentID:          token.AgentID,
+		AccessMethod:     token.AccessMethod,
+		SlugOrPort:       token.AppSlugOrPort,
+		SessionID:        uuid.New(),
+		SessionStartTime: database.Now(),
+	}
+}
+
+// StatsReporter reports workspace app StatsReports.
+type StatsReporter interface {
+	Report(context.Context, []StatsReport) error
+}
+
+// StatsDBReporter writes workspace app StatsReports to the database.
+type StatsDBReporter struct {
+	Database database.Store
+}
+
+// NewStatsDBReporter returns a new StatsDBReporter.
+func NewStatsDBReporter(db database.Store) *StatsDBReporter {
+	return &StatsDBReporter{
+		Database: db,
+	}
+}
+
+// Report writes the given StatsReports to the database.
+func (r *StatsDBReporter) Report(ctx context.Context, stats []StatsReport) error {
+	err := r.Database.InTx(func(tx database.Store) error {
+		for _, stat := range stats {
+			err := tx.InsertWorkspaceAppStats(ctx, database.InsertWorkspaceAppStatsParams{
```

Comment:
I remember than Cian refactored agent stats into batches, so probably you should adjust this part as well.

Reply:
Do you mean using sql COPY instead of INSERT? That could have an impact but if that wasn't what you meant, these are already batched by the collector so I don't think additional batching here will be necessary.

Reply:
Something similar to https://github.com/coder/coder/commit/9fb18f3ae5c0024474439cf9de6795a2d3e7a125  , so that an array of stats is unnested within a single query.

Reply:
Sure, I can implement that 👍🏻 (albeit a simpler version will do here).

Reply:
This can be done later. I just wanted to raise that we identified a similar issue in the past, and decided to fix it.

Reply:
I started working on this, but I identified a problem: it doesn't seem possible to define values that are nullable.

```sql
-- name: InsertWorkspaceAppStats :exec
INSERT INTO
	workspace_app_stats (
		id,
		user_id,
		workspace_id,
		agent_id,
		access_method,
		slug_or_port,
		session_id,
		session_started_at,
		session_ended_at
	)
SELECT
	unnest(@id::uuid[]) AS id,
	unnest(@user_id::uuid[]) AS user_id,
	unnest(@workspace_id::uuid[]) AS workspace_id,
	unnest(@agent_id::uuid[]) AS agent_id,
	unnest(@access_method::text[]) AS access_method,
	unnest(@slug_or_port::text[]) AS slug_or_port,
	unnest(@session_id::uuid[]) AS session_id,
	unnest(@session_started_at::timestamptz[]) AS session_started_at,
	unnest(@session_ended_at::nulltimestamptz[]) AS session_ended_at
ON CONFLICT
	(agent_id, session_id)
DO
	UPDATE SET
		-- Only session end can be updated.
		session_ended_at = EXCLUDED.session_ended_at
	WHERE
		workspace_app_stats.agent_id = EXCLUDED.agent_id
		AND workspace_app_stats.session_id = EXCLUDED.session_id;
```

I tried using `sqlc.narg` and also defining a new type (`nulltimestamptz`) in `sqlc.yaml`, but the former didn't work and the latter produces `interface{}`:

```yaml
      - db_type: "nulltimestamptz"
        nullable: true
        engine: "postgresql"
        go_type:
          import: "database/sql"
          type: "NullTime"
```

Since I'm refactoring the format, though, I might get away with not using nullable columns. We'll see.

Reply:
Actually, I got this `nulltimestamptz` to work... somehow. Interesting. Not sure why it worked after another try.

Reply:
As mentioned, this will still change, but: [`0bb672a` (#9001)](https://github.com/coder/coder/pull/9001/commits/0bb672ac5295d54d25c50af468305c585032fed9)